### PR TITLE
Do not escape special chars when dealing with DB strings

### PR DESCRIPTION
**Pull request by julen**
This PR implements the following changes:
* Technical details of escape sequences are omitted in the output displayed to end users
* All highlighting code has been unified on the client side.
* Special characters are depicted by symbols from a custom-built font.

Please follow-up with any comments/discussions in #3941.

### CODE REVIEW

File: pootle/static/js/utils.js [hunk 1]
```diff
@@ -78,62 +74,111 @@ var strCmp = function (a, b) {
...
-  return text.replace(escapeRE, replace);
-};
+  return text.replace(ESCAPE_RE, replace);
+}
 
 
-/* Highlight spaces to make them easily visible */
-var fancySpaces = function (text) {
+const spaceHl = '<span class="highlight-whitespace js-editor-copytext" data-string="%s"></span>';
+const WHITESPACE_RE = /^([ \u{00a0}]+)|([ \u{00a0}]+)$|[ \u{00a0}]{2,}/gmu;
```
Comment (dwaynebailey) [inline pootle/static/js/utils.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2059748190/issues/source?bid=1990595&fileBranchId=2945338#l122)

File: pootle/static/js/editor/containers/Editor.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      if (this.props.isRawMode !== undefined) {
+        extraProps.isRawMode = this.props.isRawMode;
+      }
+      editorTextareas.push(
+        <EditorTextarea
+          id={getAreaId(i)}
+          initialValue={this.props.initialValues[i]}
+          key={i}
+          onChange={(value) => this.props.onChange(i, value)}
+          {...extraProps}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/containers/Editor.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2100760359/issues/source?bid=1990595&fileBranchId=2945338#l61)

File: pootle/static/js/editor/containers/Editor.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      }
+      if (this.props.isRawMode !== undefined) {
+        extraProps.isRawMode = this.props.isRawMode;
+      }
+      editorTextareas.push(
+        <EditorTextarea
+          id={getAreaId(i)}
+          initialValue={this.props.initialValues[i]}
+          key={i}
+          onChange={(value) => this.props.onChange(i, value)}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/containers/Editor.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2100760359/issues/source?bid=1990595&fileBranchId=2945338#l60)

File: pootle/static/js/editor/containers/Editor.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        // XXX: hacky way to denote the value was already passed down
+        values[i] = null;
+      }
+      if (this.props.isRawMode !== undefined) {
+        extraProps.isRawMode = this.props.isRawMode;
+      }
+      editorTextareas.push(
+        <EditorTextarea
+          id={getAreaId(i)}
+          initialValue={this.props.initialValues[i]}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/containers/Editor.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2100760359/issues/source?bid=1990595&fileBranchId=2945338#l58)

File: pootle/static/js/editor/index.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    this.setProps(props);
+  },
+
+  setProps(props) {
+    this.props = assign(this.props, props);
+
+    ReactRenderer.render(
+      <Editor
+        onChange={(name, value) => this.handleValueChange(name, value)}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/index.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2100760360/issues/source?bid=1990595&fileBranchId=2945338#l39)

File: pootle/static/js/editor/index.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        {...this.props}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/index.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2100760360/issues/source?bid=1990595&fileBranchId=2945338#l40)

File: pootle/static/js/editor/index.js (hunk 5 above)
Comment (dwaynebailey) [inline pootle/static/js/editor/index.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2110448272/issues/source?bid=1990595&fileBranchId=2945338#l39)

File: pootle/static/js/editor/index.js (hunk 6 above)
Comment (dwaynebailey) [inline pootle/static/js/editor/index.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2110448272/issues/source?bid=1990595&fileBranchId=2945338#l40)

File: pootle/static/js/editor/containers/Editor.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        values[i] = null;
+      }
+      if (this.props.isRawMode !== undefined) {
+        extraProps.isRawMode = this.props.isRawMode;
+      }
+      editorTextareas.push(
+        <EditorTextarea
+          textareaComponent={this.props.textareaComponent}
+          id={getAreaId(i)}
+          initialValue={this.props.initialValues[i]}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/containers/Editor.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2110448271/issues/source?bid=1990595&fileBranchId=2945338#l67)

File: pootle/static/js/editor/containers/Editor.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+          key={i}
+          onChange={(value) => this.props.onChange(i, value)}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/containers/Editor.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2110448271/issues/source?bid=1990595&fileBranchId=2945338#l69)

File: pootle/static/js/editor/containers/Editor.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+          {...extraProps}
```
Comment (dwaynebailey) [inline pootle/static/js/editor/containers/Editor.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2110448271/issues/source?bid=1990595&fileBranchId=2945338#l70)

**Comment by dwaynebailey** [comment]
@julen raising this here instead of in the cloudy space of #3941 

The fact that `\r\n` isn't working and isn't being tested raises concerns for me.  I tested this, spotted brokenness and reported this fairly early in the cycle.  Yet it seems now its a surprise and we're having a discussion about `\r\n` converting to `\n`.

Can we please:

1. Not scope creep this and just deal with the one thing
2. Actually test for the escapes that are possible in localisation formats, not just the ones the Evernote has seen in its files.  That does mean `\r` on its own `\v` etc.  They're unlikely but the editor should handle them.

I'll continue other conversations on the #3941

**Comment by julen** [comment]
There's [some feedback](https://github.com/translate/pootle/issues/3941#issuecomment-215202366) re. `\r\n` which I seem to have overlooked (my bad):
* _"Three\r\n" misses CR in the source text. Target text renders fine. Edit this is when unit has editor focus, when in the context list it renders fine._
* _"Three\r\nFour" as above CR is missing in source text, target renders fine._

I spent a big part of the day debugging this and my conclusion is clear, while not new: relying on the DOM to read data will always be flawed, and this is a situation where it shines.

[Browsers adjust](https://www.w3.org/TR/CSS21/text.html#white-space-model) the white-space/line break characters they display based on different criteria. Since some of the client-side highlighting code reads the data by reading directly from the element's `textContent` attribute, this is already missing carriage return characters that might have been part of the raw data, hence causing the mismatch between parts which are using data provided directly to JS (rendered as _expected_), and those parts which read from the DOM (rendered as if there was a `\n` only).

**Comment by julen** [comment]
At the same time, there might have been some misunderstanding from my side — [I read](https://github.com/translate/pootle/issues/3941#issuecomment-157951332) _"\r\n -> LF - not sure if that was intentional but it works and makes things easier for the localiser"_ and interpreted it as a _looks good_; reading the latest comments in the issue show otherwise though.

My main feeling here is that this ideally needs to be handled transparently as you already pointed out, and at the same time, from my understanding that would probably render the above-mentioned problem as a non-issue, as users would just see a single symbol.
_(Still, I firmly believe the proper way to handle the problem mentioned in the previous comment is by reading raw data from a trusted source, and not the DOM)._

I'm in a rush now and I cannot comment on a specific way forward, but I definitely read the existing comments and hope we can make this work properly (if it's not clear, I also do believe this is a blocker).

**Comment by iafan** [comment]
Let me move the relevant part here for further discussion:

===

We can hack the editor to do the following:

1. detect if there's a `\r\n` sequence in the source string. If there is, raise the `hasCRLFEndings` flag.
2. replace all `\r\n` with `\n` before editing. Users will see linebreaks as simple `[LF]` symbols (in other words, they will be isolated from the differences in line endings, and the editor will work as expected.
3. Before submitting the translation, if `hasCRLFEndings` flag is raised, convert all lone `\n` to `\r\n`.

So in the output units, their line endings will be preserved to our best knowledge, but in Pootle people will never deal with different kinds of line endings.

===

I believe this will address the concern about handling `\r` without complicating the editor code much.

**Comment by dwaynebailey** [comment]
@iafan I think that would work.  But I'm left with a strong nagging feeling that:

1. That sound remarkably like the escaping process we've moved away from with this PR.
2. That, as you said, the correct place should most likely be in the serialisation part, not here.  Doing it here means lots of confusion to devs who can't find these strings in the UI.
3. That somehow we've not thought of all these cases and designing solutions.

I feel a lot safer if:

1. We treated `\r\n` as a unit and leaving normalization of those to another PR.
2. We knew we had full coverage for escapes that we'll hit in localisation.
3. Users were able to enter any of these escapes easily in the UI without needing to copy-paste.

**Comment by iafan** [comment]
> We treated \r\n as a unit and leaving normalization of those to another PR.

This is the same thing I'm proposing — treating internally `\r\n` as a unit. Technically this means that we need to replace `\r\n` with a single symbol, publish this in the editor as such, then convert it back. You also said you wanted consistency within a unit (not to mix `\r\n` and `\n`), so the easiest way is to do the conversion once, then allow the users work with line breaks as they do now, then do the backwards conversion once as well upon saving user input. Am I missing something?

> We knew we had full coverage for escapes that we'll hit in localisation.

Let's define what the full coverage means, then. Anything else besides `\r`, `\n` and `\t` you'd like to have supported?

> Users were able to enter any of these escapes easily in the UI without needing to copy-paste.

We can definitely render such symbols as placeables in the source string.

**Comment by dwaynebailey** [comment]
@julen I've been wondering how best to land this and how best to get wider testing.  I fear the test ask is doing simplistic testing and not deep use testing.

I have some proposals:

1. We really need unit testing here.  Without it I'm having to retest everything and we can't protect against regressions.  I think this is a good time to bring in unit testing or functional testing into our UI.
2. Deliver two editors.  If users are able to use either editor reverting to the existing one if things fail then perhaps we have a chance of getting more thorough testing before we drop the old.  Of course I'm not sure how problematic that is considering the changes that have needed to be made in the Django+JS.  But it would certainly give us a safer test driven approach to landing.

**Comment by iafan** [comment]
@dwaynebailey Julen will make unit tests that will cover bi-directional data conversion and handling all supported special symbols.

Taras also told me about your proposal for us to release this to our production server from the branch, and monitor for bugs/feedback. I'm ok with this approach, provided we have some strict cutoff date (we don't want to maintain/rebase this branch for too long). Let's do this for 2 weeks, and if there are no issues, merge the PR. If there are issues, we will fix them, and wait for another 2 weeks after any related fix). This will allow us to test this on real production data; we also have an Arabic translator who actively translates our projects, so we will ask him to provide us feedback if he spots any issues.

**Comment by dwaynebailey** [comment]
@iafan I really don't think the issue is cutoff dates and rebasing.  The central issue is actual user testing.  You seem confident in the PR, and if you are willing to do dogfood testing in a controlled space, i.e. your server, then landing it to me will happen when bugs are eliminated and enough user test coverage has happened in both RTL and LTR that we're confident we can move this into master.

**Comment by iafan** [comment]
My question here is how much testing is "enough"?

**Comment by dwaynebailey** [comment]
@iafan I think my pushback is simply that windows aren't really a way of determining if things are really working out well.  If things are very stable and the bugs are minor then it's likely too long a period, if things are very bad then it's too short.  We'll know pretty quickly how well its working and we can decide in any one of the weekly triage meetings if its good enough to land on master and coordinate a role out to our own server for testing.

Having said that, as a rough plan your idea of two week bug free sounds perfect to me. And very happy that you guys are willing to dogfood this.

**Comment by dwaynebailey** [comment]
@iafan from our side, we'll be putting up an mlo staging instance and inviting those users to work a little more like they would regularly on the editor.  My main area of concern is of course RTL, but also Thai and others.  And we have some pretty good user of LTR which would be good to run through its paces.  Using an mlo staging should make it easier to have them test and logins won't be a problem.

**Comment by iafan** [comment]
> two week bug free sounds perfect to me

Great, this is what I wanted to hear. My intent was to have some straightforward criteria on when we can land it, otherwise there will always be a temptation to test it for a few more days, and then for some more, and we don't want to sit on the branch for too long.

We will be doing a release from master tomorrow (because there were a few things changed that we want to push out first), and tentatively a week after we will do a release with this PR.

**Comment by julen** [comment]
We have run our production server off this PR branch for two weeks and so far we are happy with the state of things. The feature has been actively used by translators across the globe for locales such as Arabic, Chinese, Japanese, Korean, Thai or Vietnamese.

As a summary, we initially we experienced an important issue with IMEs but it's been quickly and properly addressed. Besides that we have kept polishing the feature and improving test coverage (and also started doing some experiments with functional testing).

Having said that, there are no critical technical blockers we are aware of, therefore we are ready to merge this PR. We already filed issues for further enhancements (#5168 and #5169), but these are not a problem of this PR specifically and are nice-to-have future improvements.

**Comment by dwaynebailey** [comment]
@julen lgtm from my side.

Landing now means this can make it into b4.  We'll do our own staging before deploying but we've got no reason to block.

**Pull request merged**